include/pybind11/cast.h

```diff
@@ -1366,6 +1366,45 @@ object object_or_cast(T &&o) {
     return pybind11::cast(std::forward<T>(o));
 }
 
+// This is being used to get around the conflict with the deprecated str() function on object_api
+using py_str = str;
+
+// Declared in pytypes.h:
+// Written here so make_caster<T> can be used
+template <typename D>
+template <typename T>
+obj_attr_accessor object_api<D>::attr_with_type_hint(handle key) const {
```

Comment:
The code duplication is far more than I'd be comfortable with. This is the kind of thing eroding quality of implementation.

Also, a test for this overload needs to be added, could you please do that first? It could be as simple as e.g.

```
    m.attr_with_type_hint<py::typing::List<int>>("list_int") = py::list(); // Exercises const char* overload.
    m.attr_with_type_hint<py::typing::Set<py::str>>(py::str("set_str")) = py::set(); // Exercises py::handle overload.
```

Re code duplication, an easy solution could be to call the `const char *` overload with `key.cast<std::string>().c_str()`. In the worst case that has a (super tiny for all practical purposes) runtime overhead. I don't think it worth the time trying to optimize that.

Reply:
Trying the following code yields a `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc0 in position 0: invalid start byte`. Do you have any suggestions on how to avoid this error?

```C++
template <typename D>
template <typename T>
str_attr_accessor object_api<D>::attr_with_type_hint(handle key) const {
    return attr_with_type_hint<T>(key.cast<std::string>().c_str());
}
```

Reply:
You changed the return type.